numpy/core/src/common/simd/sse/operators.h

```diff
@@ -0,0 +1,258 @@
+#define npyv_and_u32 _mm_and_si128
+#define npyv_and_s32 _mm_and_si128
+#define npyv_and_u64 _mm_and_si128
+#define npyv_and_s64 _mm_and_si128
+#define npyv_and_f32 _mm_and_ps
+#define npyv_and_f64 _mm_and_pd
+
+// OR
+#define npyv_or_u8  _mm_or_si128
+#define npyv_or_s8  _mm_or_si128
+#define npyv_or_u16 _mm_or_si128
+#define npyv_or_s16 _mm_or_si128
+#define npyv_or_u32 _mm_or_si128
+#define npyv_or_s32 _mm_or_si128
+#define npyv_or_u64 _mm_or_si128
+#define npyv_or_s64 _mm_or_si128
+#define npyv_or_f32 _mm_or_ps
+#define npyv_or_f64 _mm_or_pd
+
+// XOR
+#define npyv_xor_u8  _mm_xor_si128
+#define npyv_xor_s8  _mm_xor_si128
+#define npyv_xor_u16 _mm_xor_si128
+#define npyv_xor_s16 _mm_xor_si128
+#define npyv_xor_u32 _mm_xor_si128
+#define npyv_xor_s32 _mm_xor_si128
+#define npyv_xor_u64 _mm_xor_si128
+#define npyv_xor_s64 _mm_xor_si128
+#define npyv_xor_f32 _mm_xor_ps
+#define npyv_xor_f64 _mm_xor_pd
+
+// NOT
+#define npyv_not_u8(A) _mm_xor_si128(A, _mm_set1_epi32(-1))
+#define npyv_not_s8  npyv_not_u8
+#define npyv_not_u16 npyv_not_u8
+#define npyv_not_s16 npyv_not_u8
+#define npyv_not_u32 npyv_not_u8
+#define npyv_not_s32 npyv_not_u8
+#define npyv_not_u64 npyv_not_u8
+#define npyv_not_s64 npyv_not_u8
+#define npyv_not_f32(A) _mm_xor_ps(A, _mm_castsi128_ps(_mm_set1_epi32(-1)))
+#define npyv_not_f64(A) _mm_xor_pd(A, _mm_castsi128_pd(_mm_set1_epi32(-1)))
+
+/***************************
+ * Comparison
+ ***************************/
+
+// Int Equal
+#define npyv_cmpeq_u8  _mm_cmpeq_epi8
+#define npyv_cmpeq_s8  _mm_cmpeq_epi8
+#define npyv_cmpeq_u16 _mm_cmpeq_epi16
+#define npyv_cmpeq_s16 _mm_cmpeq_epi16
+#define npyv_cmpeq_u32 _mm_cmpeq_epi32
+#define npyv_cmpeq_s32 _mm_cmpeq_epi32
+#define npyv_cmpeq_s64  npyv_cmpeq_u64
+
+#ifdef NPY_HAVE_SSE41
+    #define npyv_cmpeq_u64 _mm_cmpeq_epi64
+#else
+    NPY_FINLINE __m128i npyv_cmpeq_u64(__m128i a, __m128i b)
+    {
+        __m128i cmpeq = _mm_cmpeq_epi32(a, b);
+        __m128i cmpeq_h = _mm_srli_epi64(cmpeq, 32);
+        __m128i test = _mm_and_si128(cmpeq, cmpeq_h);
+        return _mm_shuffle_epi32(test, _MM_SHUFFLE(2, 2, 0, 0));
+    }
+#endif
+
+// Int Not Equal
+#ifdef NPY_HAVE_XOP
+    #define npyv_cmpneq_u8  _mm_comneq_epi8
+    #define npyv_cmpneq_u16 _mm_comneq_epi16
+    #define npyv_cmpneq_u32 _mm_comneq_epi32
+    #define npyv_cmpneq_u64 _mm_comneq_epi64
+#else
+    #define npyv_cmpneq_u8(A, B)  npyv_not_u8(npyv_cmpeq_u8(A, B))
+    #define npyv_cmpneq_u16(A, B) npyv_not_u16(npyv_cmpeq_u16(A, B))
+    #define npyv_cmpneq_u32(A, B) npyv_not_u32(npyv_cmpeq_u32(A, B))
+    #define npyv_cmpneq_u64(A, B) npyv_not_u64(npyv_cmpeq_u64(A, B))
+#endif
+#define npyv_cmpneq_s8  npyv_cmpneq_u8
+#define npyv_cmpneq_s16 npyv_cmpneq_u16
+#define npyv_cmpneq_s32 npyv_cmpneq_u32
+#define npyv_cmpneq_s64 npyv_cmpneq_u64
+
+// signed greater than
+#define npyv_cmpgt_s8  _mm_cmpgt_epi8
+#define npyv_cmpgt_s16 _mm_cmpgt_epi16
+#define npyv_cmpgt_s32 _mm_cmpgt_epi32
+
+#ifdef NPY_HAVE_SSE42
```

Comment:
should be `NPY_HAVE_SSE41`

Reply:
No, it's correct. confusing a little bit I know but instruction `pcmpgtq` A.K.A `_mm_cmpgt_epi64` is part of SSE42, while `pcmpeqq` A.K.A `_mm_cmpeq_epi64` is part of SSE41.